crates/prover/src/preprocessed/bitwise.rs

```diff
@@ -0,0 +1,357 @@
+    ) {
+        let log_size = interaction_claim_data.bitwise.len().ilog2() + LOG_N_LANES;
+        let mut interaction_trace = LogupTraceGenerator::new(log_size);
+
+        let mut col = interaction_trace.new_col();
+        (col.par_iter_mut(), &interaction_claim_data.bitwise)
+            .into_par_iter()
+            .for_each(|(writer, value)| {
+                let denom: PackedQM31 = bitwise.combine(&[value[0], value[1], value[2], value[3]]);
+                writer.write_frac(value[4].into(), denom);
+            });
+        col.finalize_col();
+
+        let (trace, claimed_sum) = interaction_trace.finalize_last();
+        let interaction_claim = Self { claimed_sum };
+        (interaction_claim, trace)
+    }
+}
+
+#[derive(Clone)]
+pub struct Eval {
+    pub claim: Claim,
+    pub relation: crate::relations::Bitwise,
+    pub claimed_sum: SecureField,
+}
+
+impl FrameworkEval for Eval {
+    fn log_size(&self) -> u32 {
+        self.claim.log_size
+    }
+
+    fn max_constraint_log_degree_bound(&self) -> u32 {
+        self.log_size() + 1
+    }
+
+    fn evaluate<E: EvalAtRow>(&self, mut eval: E) -> E {
+        // Read the 5 trace columns
+        let operation_id =
+            eval.get_preprocessed_column(BitwiseStacked::new(0, BITWISE_OPERAND_BITS).id());
+        let input1 =
+            eval.get_preprocessed_column(BitwiseStacked::new(1, BITWISE_OPERAND_BITS).id());
+        let input2 =
+            eval.get_preprocessed_column(BitwiseStacked::new(2, BITWISE_OPERAND_BITS).id());
+        let result =
+            eval.get_preprocessed_column(BitwiseStacked::new(3, BITWISE_OPERAND_BITS).id());
+        let multiplicity = eval.next_trace_mask();
+
+        // Add lookups to the relation
+        eval.add_to_relation(RelationEntry::new(
+            &self.relation,
+            E::EF::from(multiplicity),
+            &[operation_id, input1, input2, result],
+        ));
+
+        eval.finalize_logup();
+        eval
+    }
+}
+
+pub type Component = FrameworkComponent<Eval>;
+
+/// Bitwise preprocessed columns container
+/// Represents all columns needed for bitwise operations (operation_id, input1, input2, result)
+pub struct Bitwise {
+    operand_bits: u32,
+}
+
+impl Bitwise {
+    pub const fn new(operand_bits: u32) -> Self {
+        // Could support other sizes, but for now we only use 8-bit
+        Self { operand_bits }
+    }
+
+    /// Returns the 4 columns needed for bitwise operations
+    pub const fn columns(&self) -> [BitwiseStacked; 4] {
+        [
+            BitwiseStacked::new(0, self.operand_bits),
+            BitwiseStacked::new(1, self.operand_bits),
+            BitwiseStacked::new(2, self.operand_bits),
+            BitwiseStacked::new(3, self.operand_bits),
+        ]
+    }
+}
+
+/// Stacked bitwise preprocessed column
+/// Stacks AND, OR, XOR operations into a single column
+pub struct BitwiseStacked {
```

Comment:
rename as `BitwiseCol` and remove `pub`. We don't want to see this outside of this file